File: Functions/optimize.py
import json
import sys
import requests
from scipy.optimize import linprog


# Read input data from the temporary file
if len(sys.argv) > 1:
    temp_file = sys.argv[1]
    with open(temp_file, 'r') as file:
        input_data = file.read()
    try:
        data = json.loads(input_data)
    except json.JSONDecodeError as e:
        sys.exit(1)

    min_calories = float(data['min_calories'])
    min_protein = float(data['min_protein'])
    min_fat = float(data['min_fat'])
    min_acids = float(data['min_acids'])
    min_carbs = float(data['min_carbs'])
    min_salt = float(data['min_salt'])
    min_sugar = float(data['min_sugar'])
    products = data['products']
    current_menu = data['currentMenu']  # Get currentMenu

    # Prepare data for linear programming
    c = []  # Coefficients for the objective function (prices)
    A = []  # Coefficients for the inequality constraints
    b = []  # Right-hand side values for the inequality constraints

    # Constraints: min_calories, min_fat, min_acids, min_carbs, min_salt, min_sugar
    A.append([float(product['calories']) for product in products.values()])
    b.append(min_calories)
    A.append([float(product['protein']) for product in products.values()])
    b.append(min_protein)
    A.append([float(product['fat']) for product in products.values()])
    b.append(min_fat)
    A.append([float(product['acids']) for product in products.values()])
    b.append(min_acids)
    A.append([float(product['carbs']) for product in products.values()])
    b.append(min_carbs)
    A.append([float(product['salt']) for product in products.values()])
    b.append(min_salt)
    A.append([float(product['sugar']) for product in products.values()])
    b.append(min_sugar)

    # Objective function: minimize total price
    # Convert A and b to negative for the inequality constraints (Ax >= b -> -Ax <= -b)
    A = [[-a for a in row] for row in A]
    b = [-val for val in b]
    c = [product['price'] for product in products.values()]

    # Bounds for each variable (quantity of each product)
    x_bounds = [
        (
            product['quantity'] if product.get('is_minimum', 0) == 1 else 0, 
            product['max_quantity'] if product['max_quantity'] is not None else float('inf')
        ) 
        for product in products.values()
    ]

    # Solve the linear programming problem using the Simplex method
    res = linprog(c, A_ub=A, b_ub=b, bounds=x_bounds, method='highs')
    if res.success:
        optimized_quantities = res.x
        optimized_products = []
        for i, product_id in enumerate(products.keys()):
            optimized_products.append({
                'id': product_id,
                'quantity': round(optimized_quantities[i], 2)
            })

        # Include currentMenu in the data sent to PHP script
        data_to_send = {
            'currentMenu': current_menu,
            'optimized_products': optimized_products,
            'success': True
        }

        # Redirect to menu.php with success parameters
        print(f"Location: ../menu.php?calories={round(min_calories)}&protein={round(min_protein)}&fat={round(min_fat)}&fatacids={round(min_acids)}&carb={round(min_carbs)}&salt={round(min_salt)}&sugar={round(min_sugar)}#result_section")
    else:
        data_to_send = {
            'success': False,
            'message': res.message
        }
        # Redirect to menu.php with failure message
        print("Location: ../menu.php?optimization_failed=true#result_section")

    # Send optimized data to PHP file
    url = 'http://localhost/Projektesanas_labratorija/Functions/update_quantities.php'
    headers = {'Content-Type': 'application/json'}
    try:
        response = requests.post(url, headers=headers, data=json.dumps(data_to_send))
        response.raise_for_status()  # Raise an error for bad status codes
    except requests.exceptions.RequestException as e:
        print(f"Error sending data to PHP file: {e}")

File: Functions/optimize.php
<?php
ob_start(); // Start output buffering
require '../Database/db.php';
require '../Functions/Auth.php';
checkAuth();

if (!isset($_SESSION['currentMenu'])) {
    echo "No current menu selected.";
    exit();
}

$menuid = $_SESSION['currentMenu'];

// User constraints from form submission
$minCalories = isset($_POST['min_calories']) ? $_POST['min_calories'] : 0;
$minProtein = isset($_POST['min_protein']) ? $_POST['min_protein']: 0 ;
$minFat = isset($_POST['min_fat']) ? $_POST['min_fat'] : 0;
$minAcids = isset($_POST['min_acids']) ? $_POST['min_acids'] : 0;
$minCarbs = isset($_POST['min_carbs']) ? $_POST['min_carbs'] : 0;
$minSalt = isset($_POST['min_salt']) ? $_POST['min_salt'] : 0;
$minSugar = isset($_POST['min_sugar']) ? $_POST['min_sugar'] : 0;

// Retrieve products connected to the current menu
$sql = "SELECT p.ID, p.NAME, p.CALORIES, p.PROTEIN, p.PRICE, p.SUGAR,p.ACIDS, p.FAT, p.CARBOHYDRATES, p.SALT, fp.QUANTITY, fp.MIN_QUANTITY, fp.MAX_QUANTITY
        FROM food_products fp
        INNER JOIN products p ON fp.PRODUCT_ID = p.ID
        WHERE fp.FOOD_ID = ?";
$stmt = $conn->prepare($sql);
$stmt->bind_param("i", $menuid);
$stmt->execute();
$result = $stmt->get_result();

$selectedProducts = [];
while ($row = $result->fetch_assoc()) {
    $selectedProducts[$row['ID']] = [
        'calories' => $row['CALORIES'],
        'protein' => $row['PROTEIN'],
        'price' => $row['PRICE'],
        'sugar' => $row['SUGAR'],
        'fat' => $row['FAT'],
        'carbs' => $row['CARBOHYDRATES'],
        'salt' => $row['SALT'],
        'acids' => $row['ACIDS'], // Assuming 'acids' is the same as 'fat'
        'quantity' => $row['QUANTITY'],
        'max_quantity'=>$row['MAX_QUANTITY'],
        'is_minimum'=>$row['MIN_QUANTITY']
    ];
}

// Prepare data to send to Python script
$data = [
    'min_calories' => $minCalories,
    'min_protein' => $minProtein,
    'min_fat' => $minFat,
    'min_acids' => $minAcids,
    'min_carbs' => $minCarbs,
    'min_salt' => $minSalt,
    'min_sugar' => $minSugar,
    'products' => $selectedProducts,
    'currentMenu' => $menuid 
];

// Encode data as JSON
$json_data = json_encode($data);

// Write JSON data to a temporary file
$temp_file = tempnam(sys_get_temp_dir(), 'optimize_data_');
file_put_contents($temp_file, $json_data);

// Run the Python script using the correct path to the Python executable
$python_path = "C:/Python312/python.exe";
$script_path = "C:/xampp/htdocs/Projektesanas_labratorija/Functions/optimize.py";
$command = escapeshellcmd("$python_path $script_path $temp_file");
$output = shell_exec("$command 2>&1"); // Capture both stdout and stderr

// Log the output for debugging
file_put_contents('optimize_log.txt', $output);

// Handle Python output
if ($output === NULL) {
    echo "Error running Python script.";
    ob_end_flush(); // End buffering before exiting
    exit();
}

// Check for redirection header in Python output
if (preg_match('/^Location: (.+)$/m', $output, $matches)) {
    $location = trim($matches[1]);
    header("Location: $location");
    exit();
}

echo "Error processing response from Python: $output";
ob_end_flush(); // End buffering before exiting
exit();
?>